src/metabase/task/truncate_audit_tables.clj

```diff
@@ -69,23 +69,56 @@
 Twice a day, Metabase will delete rows older than this threshold. The minimum value is 30 days (Metabase will treat entered values of 1 to 29 the same as 30).
 If set to 0, Metabase will keep all rows.")
 
+(defsetting audit-table-truncation-batch-size
+  (deferred-tru "Batch size to use for deletion of old rows for audit-related tables (like query_execution). Can be only set as an environment variable.")
+  :visibility :internal
+  :setter     :none
+  :type       :integer
+  :default    50000
+  :audit      :never
+  :export?    false)
+
+(defn- truncate-table-batched!
+  [table-name time-column]
+  (t2/query-one
+   (case (mdb/db-type)
+     (:postgres :h2)
+     {:delete-from (keyword table-name)
+      :where [:in
+              :id
+              {:select [:id]
+               :from (keyword table-name)
+               :where [:<=
+                       (keyword time-column)
+                       (t/minus (t/offset-date-time) (t/days (audit-max-retention-days)))]
+               :limit (audit-table-truncation-batch-size)}]}
+
+     (:mysql :mariadb)
+     {:delete-from (keyword table-name)
+      :where [:<=
+              (keyword time-column)
+              (t/minus (t/offset-date-time) (t/days (audit-max-retention-days)))]
+      :limit (audit-table-truncation-batch-size)})))
+
 (defn- truncate-table!
   "Given a model, deletes all rows older than the configured threshold"
   [model time-column]
   (when-not (infinite? (audit-max-retention-days))
     (let [table-name (name (t2/table-name model))]
-      (task-history/with-task-history {:task "task-history-cleanup"}
-        (try
-          (log/infof "Cleaning up %s table" table-name)
-          (let [rows-deleted (t2/delete!
-                              model
-                              time-column
-                              [:<= (t/minus (t/offset-date-time) (t/days (audit-max-retention-days)))])]
-            (if (> rows-deleted 0)
-              (log/infof "%s cleanup successful, %d rows were deleted" table-name rows-deleted)
-              (log/infof "%s cleanup successful, no rows were deleted" table-name)))
-          (catch Throwable e
-            (log/errorf e "%s cleanup failed" table-name)))))))
+      (try
+        (log/infof "Cleaning up %s table" table-name)
+        (let [total-rows-deleted (volatile! 0)]
```

Comment:
minor, but I think the speedup from `volatile!` here is unlikely to make a big difference, right? Like with the default limit of 50k rows per batch, we'd have to have 50 million rows before we executed this loop 1000 times... and it's hard to imagine that this is going to make a big impact compared to the DB round trip. So it feels a bit like something like...

```
(loop [total-deleted 0]
 (let [batch-deleted (...)]
    (if (= batch-deleted (batch-size))
      (recur (+ total-deleted batch-deleted))
      (if-not (= total-deleted 0)
        ...
```

might be clearer and more idiomatic? I'm not super tied to this though, the code you've written is pretty clear.

Reply:
That's fair, I'll change it